src/lib/swapper/swappers/ThorchainSwapper/utils/getL1ToLongtailQuote.ts

```diff
@@ -0,0 +1,174 @@
+      makeSwapErrorRight({
+        message: `[getThorTradeQuote] - No native buy asset found for ${buyChainId}.`,
+        code: TradeQuoteError.InternalError,
+        details: { buyAssetChainId: buyChainId },
+      }),
+    )
+  }
+
+  if (!nativeSellAsset) {
+    return Err(
+      makeSwapErrorRight({
+        message: `[getThorTradeQuote] - No native buy asset found for ${sellChainId}.`,
+        code: TradeQuoteError.InternalError,
+        details: { sellAssetChainId: sellChainId },
+      }),
+    )
+  }
+
+  // TODO: Move these constants outside
+  // TODO2: use more than just UniswapV3, and also consider trianglar routes.
+  const ALLOWANCE_CONTRACT = '0xF892Fef9dA200d9E84c9b0647ecFF0F34633aBe8' // TSAggregatorTokenTransferProxy
+
+  const l1Tol1QuoteInput: GetTradeQuoteInput = {
+    ...input,
+    buyAsset: nativeBuyAsset,
+    sellAsset,
+    sellAmountIncludingProtocolFeesCryptoBaseUnit: sellAmountCryptoBaseUnit,
+  }
+
+  const thorchainQuotes = await getL1quote(
+    l1Tol1QuoteInput,
+    streamingInterval,
+    TradeType.L1ToLongTail,
+  )
+
+  const unwrappedThorchainQuotes = thorchainQuotes.unwrap()
+
+  let bestAggregator: AggregatorContract
+  let quotedAmountOut: bigint
+
+  const promises = await Promise.allSettled(
+    unwrappedThorchainQuotes.map(async quote => {
+      const onlyStep = quote.steps[0]
+
+      const maybeBestAggregator = await getBestAggregator(
+        nativeBuyAsset,
+        getWrappedToken(nativeBuyAsset),
+        getTokenFromAsset(buyAsset),
+        onlyStep.buyAmountAfterFeesCryptoBaseUnit,
+      )
+
+      if (maybeBestAggregator.isErr()) return Err(maybeBestAggregator.unwrapErr())
+
+      const unwrappedResult = maybeBestAggregator.unwrap()
+
+      bestAggregator = unwrappedResult.bestAggregator
+      quotedAmountOut = unwrappedResult.quotedAmountOut
+
+      const updatedMemo = addAggregatorAndDestinationToMemo({
+        aggregator: bestAggregator,
+        destinationToken: fromAssetId(buyAsset.assetId).assetReference as Address,
+        minAmountOut: convertDecimalPercentageToBasisPoints(
+          slippageTolerancePercentageDecimal ??
+            getDefaultSlippageDecimalPercentageForSwapper(SwapperName.Thorchain),
+        ).toString(),
+        quotedMemo: quote.memo,
+      })
+
+      const { data, router, vault } = await getEvmThorTxInfo({
+        sellAsset,
+        sellAmountCryptoBaseUnit,
+        memo: updatedMemo,
+        expiry: quote.expiry,
+      })
+
+      return Ok({
+        ...quote,
+        memo: updatedMemo,
+        data,
+        router,
+        vault,
+        aggregator: bestAggregator,
+        steps: quote.steps.map(s => ({
+          ...s,
+          buyAsset,
+          buyAmountAfterFeesCryptoBaseUnit: quotedAmountOut.toString(),
+          // This is wrong, we should get the get the value before fees or display ETH value received after the thorchain bridge
+          buyAmountBeforeFeesCryptoBaseUnit: quotedAmountOut.toString(),
```

Comment:
Here, we should have the longtail token amount without gas fees and aggregator fees, but I didn't find a way to retrieve them

We have the final L1 price after the thorchain swap, but we need to have the longtail token amount before the aggregator execution, any idea except calculating it manually somehow?

Reply:
High-level, this seems like something that we'll always get wrong because of market data.

However we do support `protocolFees` as a record of asset to `ProtocolFee`, so technically, we could:

- Have the L1 fees as a first property, i.e the fees from the THOR swap
- Have the AMM fees as a second one, using whichever logic is required for it (either a contract call or an API call to get a quote from the L1 value to the destination asset if the AMM supports it)
- Use both these fees and prorate them to the buy asset, which is the part which is always doomed to be inaccurate